### PR TITLE
elf: cache crawled files

### PR DESCRIPTION
**Pull request by sergiusens**
LP: #1746607
Signed-off-by: Sergio Schvezov <sergio.schvezov@canonical.com>

- [x] Have you followed the [guidelines for contributing](https://github.com/snapcore/snapcraft/blob/master/CONTRIBUTING.md)?
- [x] Have you signed the [CLA](http://www.ubuntu.com/legal/contributors/)?
- [x] If this is a bugfix. Have you checked that there is a bug report open for the issue you are trying to fix on [bug reports](https://bugs.launchpad.net/snapcraft)?
- [x] If this is a new feature. Have you discussed the design on the [forum](https://forum.snapcraft.io)?
- [x] Have you successfully run `./runtests.sh static`?
- [x] Have you successfully run `./runtests.sh unit`?

-----

### CODE REVIEW

File: snapcraft/internal/elf.py [hunk 1]
```diff
@@ -513,6 +534,10 @@ def get_elf_files(root: str,
     """
     elf_files = set()  # type: Set[ElfFile]
 
+    # We need to reset the empty soname cache entries in case they are provided
+    # by a new set of root files
+    reset_soname_paths(only_none=True)
```
Comment (kalikiana) [inline snapcraft/internal/elf.py]: How about calling the argument `empty_entries_only`? That's the wording used in the comment and much more descriptive to my mind than referring to how it's implemented.

Comment (kalikiana) [inline snapcraft/internal/elf.py]: How about calling the argument `empty_entries_only`? That's the wording used in the comment and much more descriptive to my mind than referring to how it's implemented.

File: snapcraft/internal/elf.py [hunk 2]
```diff
@@ -89,8 +89,25 @@ def __init__(self, *, soname: str, path: str, root_path: str,
             self.in_base_snap = False
 
 
+_soname_paths = dict()  # type: Dict[str, str]
```
Comment (kyrofa) [inline snapcraft/internal/elf.py]: Global state tends to bite us. Would this be better as a class instead?

File: snapcraft/internal/elf.py (hunk 1 above)
Comment (kyrofa) [reply snapcraft/internal/elf.py]: Or consider flipping the logic. Most calls only want to remove the `None` entries. The only one that wants a complete reset is a test. Perhaps that call should be the one to include a "special" option.

**Comment by kyrofa** [comment]
Time to prime ros-talker-listener looks much better now indeed:

real	1m51.336s
user	1m23.128s
sys	0m18.288s

File: snapcraft/internal/elf.py (hunk 1 above)
Comment (sergiusens) [reply snapcraft/internal/elf.py]: sounds good.

File: snapcraft/internal/elf.py (hunk 2 above)
Comment (sergiusens) [reply snapcraft/internal/elf.py]: I can work something out

**Comment by sergiusens** [comment]
this must be a special day, adt passed on the first run, but travis did not :-P

**Comment by sergiusens** [comment]
ooh, lxd failure, showing up in all the tests on travis

**Comment by kyrofa** [comment]
> this must be a special day, adt passed on the first run, but travis did not :-P

Haha, I poked Travis like four times, too, because the error was so odd.

**Pull request merged**